### PR TITLE
CordFormation adopts new rpc security format in node.conf [CORDA-877]

### PR DESCRIPTION
**Pull request by IgorNitto**
* Cordformation now uses new format for configuring rpc access control in node configuration files    
* Added some syntactic sugar to create a list of rpc users in node dsl and used it in demos
* Modified NodeWebServer to startup with new config structure
* Incremented plugins version number

### CODE REVIEW

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt [hunk 1]
```diff
@@ -283,4 +302,15 @@ open class Cordform : DefaultTask() {
 
     private fun Node.logFile(name: String): Path = this.logDirectory().resolve(name)
     private fun ProcessBuilder.addEnvironment(key: String, value: String) = this.apply { environment().put(key, value) }
+
+    // An element of node DSL describing a user.
+    // @see listOfUsers
+    internal class NodeUser(var name: String? = null) {
```
Comment (sollecitom) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt]: Why are `username` and `password` nullable with default values?

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt [hunk 2]
```diff
@@ -283,4 +302,15 @@ open class Cordform : DefaultTask() {
...
+        var password: String? = null
+        var permissions: List<String> = mutableListOf<String>()
+        fun toPropertyMap() = mapOf(
```
Comment (sollecitom) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt]: Might be good to specify that the result type here is `Map<String, Any?>` by explicitly declared it.

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt [hunk 3]
```diff
@@ -95,7 +97,15 @@ class Node(private val project: Project) : CordformNode() {
     }
 
     private fun configureProperties() {
-        config = config.withValue("rpcUsers", ConfigValueFactory.fromIterable(rpcUsers))
+        if (rpcUsers != null) {
+            config = config.withValue("security", ConfigValueFactory.fromMap(mapOf(
```
Comment (sollecitom) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt]: Perhaps extract this into a separate private function.

File: samples/attachment-demo/build.gradle [hunk 4]
```diff
@@ -35,12 +35,18 @@ dependencies {
 }
 
 task deployNodes(type: net.corda.plugins.Cordform, dependsOn: ['jar']) {
-    ext.rpcUsers = [['username': "demo", 'password': "demo", 'permissions': ["StartFlow.net.corda.attachmentdemo.AttachmentDemoFlow",
-                                                                             "InvokeRpc.wellKnownPartyFromX500Name",
-                                                                             "InvokeRpc.attachmentExists",
-                                                                             "InvokeRpc.openAttachment",
-                                                                             "InvokeRpc.uploadAttachment",
-                                                                             "InvokeRpc.internalVerifiedTransactionsFeed"]]]
+    ext.rpcUsers = listOfUsers {
```
Comment (sollecitom) [inline samples/attachment-demo/build.gradle]: I mightily prefer this version to the current one, but why are we changing it? Do we still support the current format?

File: webserver/src/main/kotlin/net/corda/webserver/WebServerConfig.kt [hunk 5]
```diff
@@ -5,18 +5,23 @@ import net.corda.core.utilities.NetworkHostAndPort
...
 import net.corda.nodeapi.internal.config.getValue
+import net.corda.nodeapi.internal.config.parseAs
 import java.nio.file.Path
 
 /**
  * [baseDirectory] is not retrieved from the config file but rather from a command line argument.
  */
-class WebServerConfig(override val baseDirectory: Path, val config: Config) : NodeSSLConfiguration {
+class WebServerConfig(override val baseDirectory: Path, val config: Config)
+    : NodeSSLConfiguration {
```
Comment (sollecitom) [inline webserver/src/main/kotlin/net/corda/webserver/WebServerConfig.kt]: Why was this moved to new line?

File: webserver/src/main/kotlin/net/corda/webserver/WebServerConfig.kt [hunk 6]
```diff
@@ -5,18 +5,23 @@ import net.corda.core.utilities.NetworkHostAndPort
...
     override val keyStorePassword: String by config
     override val trustStorePassword: String by config
     val exportJMXto: String get() = "http"
     val useHTTPS: Boolean by config
     val myLegalName: String by config
     val rpcAddress: NetworkHostAndPort by config
     val webAddress: NetworkHostAndPort by config
-    val rpcUsers: List<User> by config
+    val runAs: User // TODO: replace with credentials supplied by a user
```
Comment (sollecitom) [inline webserver/src/main/kotlin/net/corda/webserver/WebServerConfig.kt]: Can this user come from the constructor of the configuration type? This would make it far more re-usable and testable.

File: webserver/src/main/kotlin/net/corda/webserver/internal/NodeWebServer.kt [hunk 7]
```diff
@@ -205,10 +205,9 @@ class NodeWebServer(val config: WebServerConfig) {
     }
 
     private fun connectLocalRpcAsNodeUser(): CordaRPCOps {
-        val rpcUser = config.rpcUsers.firstOrNull() ?: throw IllegalArgumentException("The node config has not specified any RPC users")
-        log.info("Connecting to node at ${config.rpcAddress} as $rpcUser")
+        log.info("Connecting to node at ${config.rpcAddress} as ${config.runAs}")
         val client = CordaRPCClient(config.rpcAddress)
-        val connection = client.start(rpcUser.username, rpcUser.password)
+        val connection = client.start(config.runAs.username, config.runAs.password)
```
Comment (sollecitom) [inline webserver/src/main/kotlin/net/corda/webserver/internal/NodeWebServer.kt]: It'd be great to have an extension function `client.startAs(user: User)` so that this line can become `val connection = client.startAs(config.user)`.

File: samples/attachment-demo/build.gradle (hunk 4 above)
Comment (sollecitom) [reply samples/attachment-demo/build.gradle]: As discussed, please leave one of the demos with the old `[[user1]]` format, to "test" that it's still compatible.

Comment (IgorNitto) [reply samples/attachment-demo/build.gradle]: The current format is still there. This is simply a macro to express the same thing with a different syntax.

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt (hunk 1 above)
Comment (IgorNitto) [reply gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt]: This is because the Gradle framework, which expect these fields to be publicly accessible AFAIK.

File: webserver/src/main/kotlin/net/corda/webserver/WebServerConfig.kt (hunk 6 above)
Comment (IgorNitto) [reply webserver/src/main/kotlin/net/corda/webserver/WebServerConfig.kt]: Sounds good. See next commit!

File: samples/attachment-demo/build.gradle (hunk 4 above)
Comment (IgorNitto) [reply samples/attachment-demo/build.gradle]: Kept plain format in previous demo.

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt [hunk 8]
```diff
@@ -54,6 +55,24 @@ open class Cordform : DefaultTask() {
...
+     *
+     * myUsers = listOfUsers {
+     *     userName1 {
+     *        password = "<password>"
+     *        permissions = ["<permission1>", ...]
+     *    }
+     *    ...
+     * }
+     */
+    fun listOfUsers(usersConfig: Closure<*>): List<Map<String, Any?>> {
```
Comment (sollecitom) [inline gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt]: Is this safe in terms of empty block or not specified?

**Comment by sollecitom** [comment]
Docs and `changelog.rst`.

File: gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt (hunk 8 above)
Comment (IgorNitto) [reply gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Cordform.kt]: I checked that an empty "listOfUsers {}" will produce an empty list of users in node.conf, while using an incorrect field inside a user block will result in a compile-time error.

File: samples/attachment-demo/build.gradle (hunk 4 above)
Comment (Clintonio) [inline samples/attachment-demo/build.gradle]: The DSL name here should be just "users". The API documentation can specify that it returns a list. This is similar to the `cordapps` block we have else where.

File: samples/irs-demo/cordapp/build.gradle [hunk 9]
```diff
@@ -45,19 +45,19 @@ dependencies {
...
-                     "StartFlow.net.corda.irs.flows.AutoOfferFlow\$Requester",
-                     "StartFlow.net.corda.irs.flows.UpdateBusinessDayFlow\$Broadcast",
-                     "StartFlow.net.corda.irs.api.NodeInterestRates\$UploadFixesFlow",
-                     "InvokeRpc.vaultQueryBy",
-                     "InvokeRpc.networkMapSnapshot",
-                     "InvokeRpc.currentNodeTime",
-                     "InvokeRpc.wellKnownPartyFromX500Name"
-             ]]
-    ]
+    ext.rpcUsers = listOfUsers {
```
Comment (Clintonio) [inline samples/irs-demo/cordapp/build.gradle]: I know you're not the original author but given we want best practices in this code can you change this to "def rpcUsers = ". Putting users on the extension makes no sense here and has been cargo culted around the codebase.

**Comment by IgorNitto** [comment]
At last, I had to revert the DSL element for rpcUsers for now because gradle plugins have changed in the meantime to a point that adding it back would be quite complicated and this has very low priority now.

The change is therefore restricted to the config output, which now uses the new format. I validated it using one of the samples to validate it.
